drivers/si1133/si1133.c

```diff
@@ -0,0 +1,353 @@
+ * depending on the configuration. 100ms is a generous limit for the HW_GAIN
+ * configuration set here.
+ */
+#define SI1133_COMMAND_TIMEOUT_USEC 100000
+
+/**
+ * @brief Run a command with no arguments.
+ *
+ * For commands CMD_PARAM_SET or CMD_PARAM_QUERY use @ref _si1133_set_param
+ * and @ref _si1133_get_param respectively instead.
+ */
+static int _si1133_run_command(si1133_t *dev, uint8_t command)
+{
+    DEBUG("[si1133] run_command: 0x%.2x, cmd_counter=%d\n",
+          (unsigned)command, dev->cmd_counter);
+    int ret;
+    ret = i2c_write_reg(dev->i2c_dev, dev->address,
+                        SI1133_REG_COMMAND, command, 0 /* flags */);
+    if (ret) {
+        DEBUG("[si1133] write command I2C error: %s\n", strerror(-ret));
+        return SI1133_ERR_I2C;
+    }
+
+    if (command == SI1133_CMD_RESET_SW) {
+        /* Reset command sets the command counter of 0x0f. */
+        dev->cmd_counter = 0x0f;
+        /* Reset command puts the device in "Initialization Mode" which requires
+         * us to wait until the device is ready. */
+        xtimer_usleep(SI1133_STARTUP_TIME);
+    }
+    else if (command == SI1133_CMD_RESET_CMD_CTR) {
+        /* The reset cmd_counter command, well, resets it to 0. */
+        dev->cmd_counter = 0;
+    }
+    else {
+        /* Increment the expected 4-bit command counter value. */
+        dev->cmd_counter = (dev->cmd_counter + 1) & SI1133_RESP0_COUNTER_MASK;
+    }
+
+    uint8_t new_cmd_ctr;
+    xtimer_ticks32_t start_time;
+    bool retry = false;
+    while (1) {
+        ret = i2c_read_reg(dev->i2c_dev, dev->address, SI1133_REG_RESPONSE0,
+                           &new_cmd_ctr, 0 /* flags */);
+        if (ret) {
+            DEBUG("[si1133] read RESPONSE0 I2C error: %s\n", strerror(-ret));
+            return SI1133_ERR_I2C;
+        }
+        if (new_cmd_ctr & SI1133_RESP0_CMD_ERR_MASK) {
+            DEBUG("[si1133] Command 0x%.2x returned error %d\n",
+                  (unsigned)command, new_cmd_ctr & SI1133_RESP0_COUNTER_MASK);
+            return SI1133_ERR_LOGIC;
+        }
+        /* The reset command is done when the RUNNING flag is clear, the other
+         * commands are done when the command value is set to the expected
+         * one. */
+        if ((command == SI1133_CMD_RESET_SW)
+            ? !(new_cmd_ctr & SI1133_RESP0_RUNNING_MASK)
+            : dev->cmd_counter
+            == (new_cmd_ctr & SI1133_RESP0_COUNTER_MASK)) {
+            break;
+        }
+        /* The command didn't yet finish in this case so it should be in running
+         * state and we need to retry the loop with a timeout. This avoids
+         * calling xtimer for commands that are immediate. */
+        if (retry) {
+            if (xtimer_usec_from_ticks(xtimer_diff(xtimer_now(), start_time))
+                > SI1133_COMMAND_TIMEOUT_USEC) {
+                DEBUG("[si1133] Command 0x%.2x timeout.\n", (unsigned)command);
+                return SI1133_ERR_LOGIC;
```

Comment:
Shouldn't the driver block until the command has finished executing?

Reply:
forever? What if the sensor hangs? should we hang our CPU in that case? The sensor's datasheet states the duration of the forced sampling based on the number of channels and the HW_GAIN / SW_GAIN parameters. I think the worst in the current setting is < 2ms, but maybe we can set higher gain in the future. If the sensor takes more than that there there's a logic error somewhere (and we should probably issue a reset / call si1133_init() again).

Reply:
I'd use a `mutex_lock_timeout()` and unlock the mutex from within the sensor's isr.

Reply:
the time is 1~2ms here if you enable all 6 sensors, so there's some potential to do something else in between with the core at 16 MHz. There's a way to get an interrupt from the sensor but only if you wire that pin to a GPIO, which you may not have available. This commit was more about the minimum possible to get this to work and does not set up the interrupt handler. Also the interrupt pin could be shared with other sensors which makes things more interesting :).

We can either:
 1.  provide a non-blocking mode where you have to call back to check if it is done (for example when the interrupt triggers).
 2. do some match to compute the total sampling time for all active channels and do a blocking wait instead of the busy loop here (allowing other threads to run in between).

Reply:
Ah I thought you had an option to configure the interrupt in here already - must have mixed that up.

Anyway, polling until timeout is fine if we don't have an interrupt source yet, I just missed that loop on my first review :sweat_smile: